File: swift/Sources/MLXAudioModels/HTDemucs/Inference.swift
// Inference.swift
// Inference utilities for HTDemucs.

import Foundation
@preconcurrency import MLX

// MARK: - Model Protocol

/// Protocol for models that can perform source separation.
public protocol SeparationModel {
    var config: HTDemucsConfig { get }
    func callAsFunction(_ mix: MLXArray) -> MLXArray
}

extension HTDemucs: SeparationModel {}
extension BagOfModels: SeparationModel {}

// MARK: - Inference

/// Apply HTDemucs model with chunked inference.
///
/// For long audio, the input is split into overlapping segments,
/// processed independently, and blended using overlap-add.
///
/// Works with both single HTDemucs models and BagOfModels ensembles.
///
/// - Parameters:
///   - model: HTDemucs model or BagOfModels ensemble.
///   - mix: Input mixture `[C, T]` or `[B, C, T]`.
///   - segment: Segment duration in seconds (nil = use model default).
///   - overlap: Overlap ratio between segments.
///   - split: Enable chunking (false = process entire audio at once).
///   - progressCallback: Optional progress callback function.
/// - Returns: Separated stems `[S, C, T]` or `[B, S, C, T]`.
public func applyModel(
    _ model: SeparationModel,
    mix: MLXArray,
    segment: Float? = nil,
    overlap: Float = 0.25,
    split: Bool = true,
    progressCallback: ((Float) -> Void)? = nil
) -> MLXArray {
    var input = mix
    var squeezeBatch = false

    // Add batch dimension if needed
    if input.ndim == 2 {
        input = input.expandedDimensions(axis: 0)
        squeezeBatch = true
    }

    let T = input.shape[2]
    let segmentValue = segment ?? model.config.segment
    let segmentSamples = Int(segmentValue * Float(model.config.samplerate))

    var stems: MLXArray

    if !split || T <= segmentSamples {
        // Process in single pass
        stems = model.callAsFunction(input)
    } else {
        // Chunked processing with overlap-add
        stems = chunkedInference(
            model: model,
            mix: input,
            segmentSamples: segmentSamples,
            overlap: overlap,
            progressCallback: progressCallback
        )
    }

    if squeezeBatch {
        stems = stems.squeezed(axis: 0)
    }

    return stems
}

/// Process long audio in overlapping chunks.
///
/// Uses overlap-add blending strategy to ensure smooth transitions
/// between chunks.
///
/// - Parameters:
///   - model: HTDemucs model or ensemble.
///   - mix: Input mixture `[B, C, T]`.
///   - segmentSamples: Samples per segment.
///   - overlap: Overlap ratio.
///   - progressCallback: Optional progress callback.
/// - Returns: Separated stems `[B, S, C, T]`.
private func chunkedInference(
    model: SeparationModel,
    mix: MLXArray,
    segmentSamples: Int,
    overlap: Float,
    progressCallback: ((Float) -> Void)?
) -> MLXArray {
    let B = mix.shape[0]
    let C = mix.shape[1]
    let T = mix.shape[2]
    let S = model.config.num_sources

    let overlapSamples = Int(Float(segmentSamples) * overlap)
    let stride = segmentSamples - overlapSamples

    // Output buffer
    var out = MLXArray.zeros([B, S, C, T])
    var weightSum = MLXArray.zeros([B, S, C, T])

    // Triangular window for overlap-add
    let weight = createWeightWindow(length: segmentSamples)

    // Calculate number of chunks
    let numChunks: Int
    if T <= segmentSamples {
        numChunks = 1
    } else {
        numChunks = (T - overlapSamples + stride - 1) / stride
    }

    for chunkIdx in 0..<numChunks {
        let offset = chunkIdx * stride
        let chunkEnd = min(offset + segmentSamples, T)
        let chunkLen = chunkEnd - offset

        // Extract chunk
        var chunk = mix[0..., 0..., offset..<chunkEnd]

        // Pad if needed
        if chunkLen < segmentSamples {
            let padAmount = segmentSamples - chunkLen
            chunk = MLX.padded(chunk, widths: [[0, 0], [0, 0], [0, padAmount]])
        }

        // Apply model
        var chunkOut = model.callAsFunction(chunk)

        // Trim to actual length if we padded
        var chunkWeight: MLXArray
        if chunkLen < segmentSamples {
            chunkOut = chunkOut[0..., 0..., 0..., 0..<chunkLen]
            chunkWeight = weight[0..<chunkLen]
        } else {
            chunkWeight = weight
        }

        // Reshape weight for broadcasting: [1, 1, 1, chunkLen]
        let chunkWeight4D = chunkWeight.reshaped([1, 1, 1, -1])

        // Weighted accumulation
        let weightedChunk = chunkOut * chunkWeight4D

        // Accumulate into output buffers
        // Note: MLX Swift may not have .at().add() - use workaround
        let slicedOut = out[0..., 0..., 0..., offset..<chunkEnd]
        let newSlice = slicedOut + weightedChunk

        // Reconstruct output with updated slice
        if offset == 0 {
            let rest = out[0..., 0..., 0..., chunkEnd...]
            out = concatenated([newSlice, rest], axis: 3)
        } else if chunkEnd == T {
            let prefix = out[0..., 0..., 0..., 0..<offset]
            out = concatenated([prefix, newSlice], axis: 3)
        } else {
            let prefix = out[0..., 0..., 0..., 0..<offset]
            let suffix = out[0..., 0..., 0..., chunkEnd...]
            out = concatenated([prefix, newSlice, suffix], axis: 3)
        }

        // Same for weight sum
        let broadcastWeight = MLX.broadcast(chunkWeight4D, to: weightedChunk.shape)
        let slicedWeight = weightSum[0..., 0..., 0..., offset..<chunkEnd]
        let newWeightSlice = slicedWeight + broadcastWeight

        if offset == 0 {
            let rest = weightSum[0..., 0..., 0..., chunkEnd...]
            weightSum = concatenated([newWeightSlice, rest], axis: 3)
        } else if chunkEnd == T {
            let prefix = weightSum[0..., 0..., 0..., 0..<offset]
            weightSum = concatenated([prefix, newWeightSlice], axis: 3)
        } else {
            let prefix = weightSum[0..., 0..., 0..., 0..<offset]
            let suffix = weightSum[0..., 0..., 0..., chunkEnd...]
            weightSum = concatenated([prefix, newWeightSlice, suffix], axis: 3)
        }

        // Progress callback
        progressCallback?(Float(chunkIdx + 1) / Float(numChunks))

        // Evaluate to avoid memory buildup
        eval(out, weightSum)
    }

    // Normalize by weights
    out = out / maximum(weightSum, MLXArray(1e-8))

    return out
}

/// Create triangular weight window for overlap-add.
///
/// - Parameters:
///   - length: Window length.
///   - power: Power to raise the window to.
/// - Returns: Weight window.
private func createWeightWindow(length: Int, power: Float = 1.0) -> MLXArray {
    let half = length / 2

    // Ramp up from 1 to half
    let rampUp = MLXArray(1...half).asType(.float32)

    // Ramp down from (length - half) to 1
    let rampDownValues = stride(from: length - half, through: 1, by: -1).map { Float($0) }
    let rampDown = MLXArray(rampDownValues)

    var weight = concatenated([rampUp, rampDown], axis: 0)
    weight = pow(weight / weight.max(), power)

    return weight
}

File: python/mlx_audio/models/demucs/inference.py
"""Inference utilities for HTDemucs."""

from __future__ import annotations

from collections.abc import Callable
from typing import TYPE_CHECKING, Any, Union

import mlx.core as mx

from mlx_audio.models.demucs.model import HTDemucs

if TYPE_CHECKING:
    from mlx_audio.models.demucs.bag import BagOfModels

# Type alias for model parameter
ModelType = Union[HTDemucs, "BagOfModels"]


def apply_model(
    model: ModelType,
    mix: mx.array,
    *,
    segment: float | None = None,
    overlap: float = 0.25,
    shifts: int = 1,
    split: bool = True,
    progress_callback: Callable[[float], None] | None = None,
    **kwargs: Any,
) -> mx.array:
    """Apply HTDemucs model with chunked inference.

    For long audio, the input is split into overlapping segments,
    processed independently, and blended using overlap-add.

    Works with both single HTDemucs models and BagOfModels ensembles.

    Args:
        model: HTDemucs model or BagOfModels ensemble
        mix: Input mixture [C, T] or [B, C, T]
        segment: Segment duration in seconds (None = use model default)
        overlap: Overlap ratio between segments
        shifts: Number of random time shifts for augmentation
        split: Enable chunking (False = process entire audio at once)
        progress_callback: Optional progress callback function
        **kwargs: Additional arguments

    Returns:
        Separated stems [S, C, T] or [B, S, C, T]
    """
    # Add batch dimension if needed
    if mix.ndim == 2:
        mix = mix[None, ...]
        squeeze_batch = True
    else:
        squeeze_batch = False

    B, C, T = mix.shape
    _ = B, C  # Unused but documents expected shape

    if segment is None:
        segment = model.config.segment

    segment_samples = int(segment * model.config.samplerate)

    if not split or segment_samples >= T:
        # Process in single pass
        stems = model(mix)
    else:
        # Chunked processing with overlap-add
        stems = _chunked_inference(
            model,
            mix,
            segment_samples=segment_samples,
            overlap=overlap,
            progress_callback=progress_callback,
        )

    if squeeze_batch:
        stems = stems[0]

    return stems


def _chunked_inference(
    model: HTDemucs,
    mix: mx.array,
    *,
    segment_samples: int,
    overlap: float,
    progress_callback: Callable[[float], None] | None,
) -> mx.array:
    """Process long audio in overlapping chunks.

    Uses the same overlap-add blending strategy as the streaming processor
    to ensure identical outputs between batch and streaming inference.

    Args:
        model: HTDemucs model
        mix: Input mixture [B, C, T]
        segment_samples: Samples per segment
        overlap: Overlap ratio
        progress_callback: Optional progress callback

    Returns:
        Separated stems [B, S, C, T]
    """
    B, C, T = mix.shape
    S = model.config.num_sources

    overlap_samples = int(segment_samples * overlap)
    stride = segment_samples - overlap_samples

    # Output buffer
    out = mx.zeros((B, S, C, T))
    weight_sum = mx.zeros((B, S, C, T))

    # Triangular window for overlap-add
    weight = _create_weight_window(segment_samples)

    # Calculate number of chunks - ensure we cover all samples
    # The last chunk may extend past T but we'll handle padding
    num_chunks = 1 if segment_samples >= T else (T - overlap_samples + stride - 1) // stride

    for chunk_idx in range(num_chunks):
        offset = chunk_idx * stride
        chunk_end = min(offset + segment_samples, T)
        chunk_len = chunk_end - offset

        # Extract chunk
        chunk = mix[:, :, offset:chunk_end]

        # Pad if needed
        if chunk_len < segment_samples:
            pad_amount = segment_samples - chunk_len
            chunk = mx.pad(chunk, [(0, 0), (0, 0), (0, pad_amount)])

        # Apply model
        chunk_out = model(chunk)

        # Trim to actual length if we padded
        if chunk_len < segment_samples:
            chunk_out = chunk_out[:, :, :, :chunk_len]
            chunk_weight = weight[:chunk_len]
        else:
            chunk_weight = weight

        # Reshape weight for broadcasting: [1, 1, 1, chunk_len]
        chunk_weight_4d = chunk_weight.reshape(1, 1, 1, -1)

        # Weighted accumulation
        weighted_chunk = chunk_out * chunk_weight_4d

        # Use slice assignment for accumulation
        out = out.at[:, :, :, offset:chunk_end].add(weighted_chunk)
        weight_sum = weight_sum.at[:, :, :, offset:chunk_end].add(
            mx.broadcast_to(chunk_weight_4d, weighted_chunk.shape)
        )

        # Progress callback
        if progress_callback:
            progress_callback((chunk_idx + 1) / num_chunks)

        # Evaluate to avoid memory buildup
        mx.eval(out, weight_sum)

    # Normalize by weights
    out = out / mx.maximum(weight_sum, 1e-8)

    return out


def _create_weight_window(length: int, power: float = 1.0) -> mx.array:
    """Create triangular weight window for overlap-add.

    Args:
        length: Window length
        power: Power to raise the window to

    Returns:
        Weight window
    """
    half = length // 2
    ramp_up = mx.arange(1, half + 1, dtype=mx.float32)
    ramp_down = mx.arange(length - half, 0, -1, dtype=mx.float32)
    weight = mx.concatenate([ramp_up, ramp_down])
    weight = (weight / mx.max(weight)) ** power
    return weight
